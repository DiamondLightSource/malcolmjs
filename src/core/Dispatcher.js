/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
