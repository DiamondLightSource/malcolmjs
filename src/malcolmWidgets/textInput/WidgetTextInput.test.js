import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import WidgetTextInput from './WidgetTextInput.component';

const textInput = (
  value,
  pending,
  units,
  isDirty,
  setFlag = () => {},
  setValue = () => {},
  forceUpdate = false,
  error = false
) => (
  <WidgetTextInput
    Value={value}
    Pending={pending}
    submitEventHandler={setValue}
    focusHandler={() => {}}
    blurHandler={() => {}}
    Units={units}
    isDirty={isDirty}
    setFlag={setFlag}
    forceUpdate={forceUpdate}
    Error={error}
  />
);

describe('WidgetTextUpdate', () => {
  let shallow;
  let mount;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    mount = createMount();
  });

  it('displays text', () => {
    const wrapper = shallow(textInput('Hello World', false, null));
    expect(wrapper).toMatchSnapshot();
  });

  it('displays with units', () => {
    const wrapper = shallow(textInput('1.21', false, 'GW'));
    expect(wrapper).toMatchSnapshot();
  });

  it('displays disabled', () => {
    const wrapper = shallow(textInput('Goodbye World', true, null));
    expect(wrapper).toMatchSnapshot();
  });

  it('displays dirty', () => {
    const wrapper = shallow(textInput('Goodbye World', true, null, true));
    expect(wrapper).toMatchSnapshot();
  });

  it('calls set dirty on focus', () => {
    const setDirty = jest.fn();
    const wrapper = mount(
      textInput('Hello World', false, null, false, setDirty)
    );
    wrapper
      .find('input')
      .first()
      .simulate('focus');
    expect(setDirty.mock.calls.length).toEqual(1);
    expect(setDirty.mock.calls[0]).toEqual(['dirty', true]);
  });

  it('calls unset dirty on focus if value unchanged', () => {
    const setDirty = jest.fn();
    const wrapper = mount(
      textInput('Hello World', false, null, false, setDirty)
    );
    wrapper
      .find('input')
      .first()
      .simulate('blur');
    expect(setDirty.mock.calls.length).toEqual(1);
    expect(setDirty.mock.calls[0]).toEqual(['dirty', false]);
  });

  it('doesnt call unset dirty on focus if value changed', () => {
    const setDirty = jest.fn();
    const wrapper = mount(
      textInput('Hello World', false, null, false, setDirty)
    );
    wrapper
      .find('input')
      .first()
      .simulate('change', { target: { value: 'test' } })
      .simulate('blur');
    expect(setDirty.mock.calls.length).toEqual(1);
    expect(setDirty.mock.calls[0]).toEqual(['dirty', true]);
  });

  it('calls send on enter', () => {
    const mockData = { mock: { calls: [] } };
    const setValue = event => {
      mockData.mock.calls.push(event.target.value);
    };
    const wrapper = mount(
      textInput('Hello World', false, null, false, () => {}, setValue)
    );
    wrapper
      .find('input')
      .first()
      .simulate('keyPress', { key: 'Enter' });
    expect(mockData.mock.calls.length).toEqual(1);
    expect(mockData.mock.calls[0]).toEqual('Hello World');
  });

  it('updates and calls unset flag if forceUpdate is true and is dirty', () => {
    const setFlag = jest.fn();
    mount(textInput('Hello World', false, null, true, setFlag, () => {}, true));
    expect(setFlag.mock.calls.length).toEqual(1);
    expect(setFlag.mock.calls[0]).toEqual(['forceUpdate', false]);
  });

  it('updates, calls unset flag and sends put with last known malcolm value if forceUpdate is true and is error', () => {
    const setFlag = jest.fn();
    const setValue = jest.fn();
    mount(
      textInput('Hello World', false, null, true, setFlag, setValue, true, true)
    );
    expect(setFlag.mock.calls.length).toEqual(1);
    expect(setFlag.mock.calls[0]).toEqual(['forceUpdate', false]);
    expect(setValue.mock.calls.length).toEqual(1);
    expect(setValue.mock.calls[0]).toEqual([
      {
        target: { value: 'Hello World' },
      },
    ]);
  });
});
