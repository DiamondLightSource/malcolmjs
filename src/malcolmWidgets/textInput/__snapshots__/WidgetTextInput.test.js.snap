// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetTextUpdate displays dirty 1`] = `
<TextField
  InputProps={
    Object {
      "className": "WidgetTextInput-InputStyle-3",
      "endAdornment": null,
    }
  }
  className="WidgetTextInput-textInput-1"
  disabled={true}
  error={false}
  inputProps={
    Object {
      "className": "WidgetTextInput-inputStyle-2",
    }
  }
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  required={false}
  select={false}
  value="Goodbye World"
/>
`;

exports[`WidgetTextUpdate displays disabled 1`] = `
<TextField
  InputProps={
    Object {
      "className": "WidgetTextInput-InputStyle-3",
      "endAdornment": null,
    }
  }
  className="WidgetTextInput-textInput-1"
  disabled={true}
  error={false}
  inputProps={
    Object {
      "className": "WidgetTextInput-inputStyle-2",
    }
  }
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  required={false}
  select={false}
  value="Goodbye World"
/>
`;

exports[`WidgetTextUpdate displays text 1`] = `
<TextField
  InputProps={
    Object {
      "className": "WidgetTextInput-InputStyle-3",
      "endAdornment": null,
    }
  }
  className="WidgetTextInput-textInput-1"
  disabled={false}
  error={false}
  inputProps={
    Object {
      "className": "WidgetTextInput-inputStyle-2",
    }
  }
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  required={false}
  select={false}
  value="Hello World"
/>
`;

exports[`WidgetTextUpdate displays with units 1`] = `
<TextField
  InputProps={
    Object {
      "className": "WidgetTextInput-InputStyle-3",
      "endAdornment": <WithStyles(InputAdornment)
        position="end"
    >
        GW
    </WithStyles(InputAdornment)>,
    }
  }
  className="WidgetTextInput-textInput-1"
  disabled={false}
  error={false}
  inputProps={
    Object {
      "className": "WidgetTextInput-inputStyle-2",
    }
  }
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  required={false}
  select={false}
  value="1.21"
/>
`;
